Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm

```diff
@@ -906,6 +907,90 @@ static bool needsEAGLOnMac()
 }
 
+void GraphicsContextGLCocoa::paintRenderingResultsToCanvas(ImageBuffer& imageBuffer)
+{
+    withDrawingBufferAsPlatformImage([&](PlatformImagePtr&& image) {
+        paintToCanvas(WTFMove(image), imageBuffer.backendSize(), imageBuffer.context());
+    });
+}
+
+void GraphicsContextGLCocoa::paintCompositedResultsToCanvas(ImageBuffer& imageBuffer)
+{
+    withDisplayBufferAsPlatformImage([&](PlatformImagePtr&& image) {
+        paintToCanvas(WTFMove(image), imageBuffer.backendSize(), imageBuffer.context());
+    });
+}
+
+void GraphicsContextGLCocoa::withDrawingBufferAsPlatformImage(std::function<void(PlatformImagePtr&&)> func)
+{
+    if (!makeContextCurrent())
+        return;
+
+    if (!m_displayBufferBacking
+        || m_displayBufferBacking->size() != getInternalFramebufferSize())
+        return;
+
+    prepareTexture();
+
+    if (contextAttributes().premultipliedAlpha) {
+        // Use the IOSurface backed image directly
+
+        GL_ScheduleANGLE();
+
+        // FIXME: Should a displayId be passed here?
+        auto cgContext = m_displayBufferBacking->createPlatformContext();
+        if (cgContext) {
+            // FIXME: Should this be CGIOSurfaceContextCreateImageReference to avoid
+            // creating a copy?
+
+            auto drawingImage = m_displayBufferBacking->createImage(cgContext.get());
+            func(WTFMove(drawingImage));
+
+            // FIXME: Should we be interacting with ImageBufferIOSurfaceBackend
+            // directly? This code is
+            // ImageBufferIOSurfaceBackend::prepareForExternalRead(), but that's
+            CGContextFlush(cgContext.get());
+        }
+    } else {
+        // Since IOSurface-backed images can only support premultiplied alpha,
+        // read the image into a PixelBuffer which can be we used to create a
+        // CGImage.
+        //
+        // FIXME: Can the IOSurface be read into a buffer to avoid the read back via GL?
+        auto drawingPixelBuffer = paintRenderingResultsToPixelBuffer();
+        if (!drawingPixelBuffer)
+            return;
+
+        auto drawingImage = createImageFromPixelBuffer(drawingPixelBuffer.releaseNonNull(), AlphaMode::Unmultiplied);
+        func(WTFMove(drawingImage));
+    }
+}
+
+void GraphicsContextGLCocoa::withDisplayBufferAsPlatformImage(std::function<void(PlatformImagePtr&&)> func)
+{
+    auto& displayBuffer = m_swapChain.displayBuffer();
+    if (!displayBuffer.surface || !displayBuffer.handle)
+        return;
+    if (displayBuffer.surface->size() != getInternalFramebufferSize())
+        return;
+
+    // FIXME: Should a displayID be passed here?
+    auto cgContext = displayBuffer.surface->createPlatformContext();
+    if (!cgContext)
+        return;
+
+    // FIXME: Should this be CGIOSurfaceContextCreateImageReference to avoid
+    // creating a copy?
+    func(displayBuffer.surface->createImage(cgContext.get()));
+
+    // FIXME: Should we be interacting with ImageBufferIOSurfaceBackend
```

Comment:
so here you don't need to flush the context used for the **source image**, as  that does not have any drawing.
Instead, you could do `CGContextFillRect(cgContext.get(), CGRect { });`.
This would ensure that if the func() hangs on to the image, we would not continue updating the image behind the func()s back.